api/server/router/system/system_routes.go

```diff
@@ -44,6 +47,28 @@ func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r
 		w.Header().Set("Content-Length", "0")
 		return nil
 	}
+
+	if capabilitiesQueryValue := r.URL.Query().Get("capabilities"); capabilitiesQueryValue != "" {
+		wantedCapabilitiesVersion, err := strconv.Atoi(capabilitiesQueryValue)
+		if err != nil {
+			return errdefs.InvalidParameter(err)
+		}
+		capabilities, err := s.handleCapabilities(ctx, wantedCapabilitiesVersion)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("failed")
+		} else {
+			w.Header().Set("Content-Type", "application/json")
+			capabilitiesJson, err := json.Marshal(types.PingBody{
+				Capabilities: &capabilities,
+			})
+			if err != nil {
+				return err
+			}
+			_, err = w.Write(capabilitiesJson)
+			return err
```

Comment:
This can probably use httputils.WriteJSON, which also handles setting the content-type etc;
https://github.com/moby/moby/blob/1e033e2d983df0f20caa8da5320dee3c3c2d7d63/api/server/router/container/container_routes.go#L70-L74